pOCCI/pOCCI.1
.TH POCCI 1 "August 2015" "CESNET" "pOCCI"


.SH NAME
pOCCI \- OCCI compliance testing and client


.SH SYNOPSIS
\fBpOCCI\fR [\fIOPTIONS\fR]


.SH DESCRIPTION
\fBpOCCI\fR is OCCI implementation written in python.

\fBpOCCI\fR tool performs OCCI compliance testing as described in:

\fIhttp://www.etsi.org/deliver/etsi_ts/103100_103199/103142/01.01.01_60/ts_103142v010101p.pdf\fR.


.SH OPTIONS
.TP
\fB-h\fR, \fP--help\fR
Usage message.

.TP
\fB-a\fR, \fP--auth-type\fR \fIAUTH-TYPE\fR
Authentication type to use. Default is \fBbasic\fR.

Available values: \fBbasic\fR, \fBx509\fR.

.TP
\fB-c\fR, \fP--cert\fR \fIFILE\fR
User SSL certificate file, if \fBx509\fR autentization is used.

For VOMS proxy certificates, it should be used the variable $X509_USER_PROXY instead.

.TP
\fB-e\fR, \fP--endpoint\fR, \fP--url\fR \fIURL\fR
OCCI server endpoint. For example: \fIhttps://example.com:11443\fR.

.TP
\fB-f\fR, \fP--format\fR \fIFORMAT\fR
Output format (\fBplain\fR, \fBjson\fR, \fBhtml\fR, \fBhtmltable\fR). Default is \fBjson\fR.

.TP
\fB-k\fR, \fP--key\fR \fIFILE\fR
User SSL key file, if \fBx509\fR autentization is used.

.TP
\fB-l\fR, \fP--list\fR
Get list of all OCCI compliance tests.

.TP
\fB-m\fR, \fP--mime-type\fR \fIMIME-TYPE\fR
Render format to communicate with OCCI server. If \fBtext/occi\fR is selected, listing of categories is performed through \fBtext/plain\fR due to HTTP Headers size limitations. Default is \fBtext/plain\fR.

Possible values: \fBtext/plain\fR, \fBtext/occi\fR, \fBtext/occi+plain\fR.

.TP
\fB-p\fR, \fP--password\fR \fIPASSWORD\fR
User password, if \fBbasic\fR autentization is used.

.TP
\fB-P\fR, \fP--passphrase\fR \fIPASSPHRASE\fR
Passphrase for SSL user key, if \fBx509\fR autentization is used. Default is no passphrase.

.TP
\fB-s\fR, \fP--keystone\fR \fITENANT\fR
Keystone tenant for authentication. For example: \fIVO:fedcloud.egi.eu\fR.

.TP
\fB-S\fR, \fP--ignore-ssl\fR
Ignore SSL errors. Default is false.

.TP
\fB-t\fR, \fP--tests\fR \fITEST1,...\fR
List of OCCI compliance tests to perform. Separated by commas.

.TP
\fB-u\fR, \fP--user\fR \fIUSER\fR
User name, if \fBbasic\fR autentization is used.

.TP
\fB-v\fR, \fP--verbose\fR
Verbose mode: communication, HTTP Headers, transferred data.

.TP
\fB-V\fR, \fP--version\fR
Print version information.

.SH EXIT STATUS
.TP
0
All OCCI compliance tests \fBOK\fR.

.TP
1
Some OCCI compliance tests \fBFAILED\fR.

.TP
2
Error from OCCI testsuite framework.


.SH ENVIRONMENT

.TP
\fBX509_CERT_DIR\fR
CA path to use.

.TP
\fBX509_USER_PROXY\fR
User proxy certificate file to use, if \fB\-\-cert\fR option is not specified.


.SH FILES

.TP
\fB~/.pOCCI.cfg\fR, \fB/etc/pOCCI.cfg\fR
pOCCI configuration file.

Available options and default values:
 * \fBurl\fR
 * \fBauthtype\fR = 'basic'
 * \fBcapath\fR = '/etc/grid-security/certificates'
 * \fBcachain\fR
 * \fBignoressl\fR = False
 * \fBuser\fR
 * \fBpasswd\fR
 * \fBcert\fR
 * \fBkey\fR
 * \fBpassphrase\fR
 * \fBkeystone\fR
 * \fBcurlverbose\fR = False
 * \fBconnectiontimeout\fR = 60
 * \fBtimeout\fR = 120
 * \fBmimetype\fR = 'text/plain'
 * \fBoutputformat\fR = 'json'
 * \fBocci.tests.category\fR = {'term': 'compute', 'class': 'kind', 'scheme': 'http://schemas.ogf.org/occi/infrastructure#'}
 * \fBocci.tests.entity\fR = {}


.SH EXAMPLES

.TP
\fBpOCCI.cfg\fR file example:

.nf
[main]
url = 'https://example.com:11443'
authtype = 'basic'
ignoressl = False
user = 'oneadmin'
passwd = 'good-password'
mimetype = 'text/plain'
outputformat = 'json'
curlverbose = False
connectiontimeout = 60
timeout = 120
occi.tests.entity = {'occi.core.title': 'one-172'}
.fi


.SH BUGS
Please report all bugs to issue tracker at \fIhttps://github.com/CESNET/pOCCI/issues\fR.


.SH AUTHORS
CESNET


pOCCI/pOCCI.py
#!/usr/bin/python

import getopt
import re
import sys
import time
import os

from occi_libs import *
from CORE import *
import occi
import version


tests_definitions = {
    'OCCI/CORE/DISCOVERY/001': CORE_DISCOVERY001,
    'OCCI/CORE/DISCOVERY/002': CORE_DISCOVERY002,
    'OCCI/CORE/READ/001': CORE_READ001,
    'OCCI/CORE/READ/002': CORE_READ002,
    'OCCI/CORE/READ/007': CORE_READ007,
    'OCCI/CORE/CREATE/001': CORE_CREATE001,
    'OCCI/CORE/CREATE/006': CORE_CREATE006,
    'OCCI/CORE/DELETE/001': CORE_DELETE001,
    'OCCI/CORE/UPDATE/001': CORE_UPDATE001,
    'OCCI/CORE/MISC/001': CORE_MISC001,

    'OCCI/INFRA/CREATE/001': INFRA_CREATE001,
    'OCCI/INFRA/CREATE/002': INFRA_CREATE002,
    'OCCI/INFRA/CREATE/003': INFRA_CREATE003,
    'OCCI/INFRA/CREATE/004': INFRA_CREATE004,
    'OCCI/INFRA/CREATE/005': INFRA_CREATE005,
    'OCCI/INFRA/CREATE/006': INFRA_CREATE006,
    'OCCI/INFRA/CREATE/007': INFRA_CREATE007,
}


tests_skipped = set([
    'OCCI/CORE/CREATE/001',
    'OCCI/CORE/CREATE/006',
    'OCCI/INFRA/CREATE/005',
])


def usage(name=__file__):
    print "%s [OPTIONS]\n\
\n\
OPTIONS:\n\
  -h, --help ................ usage message\n\
  -a, --auth-type AUTH ...... authentization type\n\
  -c, --cert FILE ........... SSL certificate file\n\
  -e, --endpoint, --url URL . OCCI server endpoint\n\
  -f, --format FORMAT ....... output format (plain, json)\n\
  -k, --key FILE ............ SSL key file\n\
  -l, --list ................ list all test\n\
  -m, --mime-type MIME-TYPE . render format\n\
  -p, --password PWD ........ password for basic auth-type\n\
  -P, --passphrase PASS ..... SSL key passphrase\n\
  -s, --keystone TENANT ..... tenant used for keystone auth\n\
  -S, --ignore-ssl .......... ignore SSL errors\n\
  -t, --tests <TEST1,...> ... list of tests\n\
  -u, --user USER ........... user name for basic auth-type\n\
  -v, --verbose ............. verbose mode\n\
  -V, --version ............. print version information\n\
\n\
MIME-TYPE: text/plain, text/occi\n\
AUTH: basic\n\
" % os.path.basename(name)


def main(argv=sys.argv[1:]):
    results = []
    tests = []

    if not occi_config:
        print >> sys.stderr, 'No configuration found'
        sys.exit(2)

    try:
        opts, args = getopt.getopt(argv, "ha:c:e:f:k:lm:p:Ps::St:u:vV", ["help", "auth-type=", "cert=", "endpoint=", "format=", "ignore-ssl", "keystone=", "key=", "list", "mime-type=", "passphrase=", "password=", "tests=", "url=", "user=", "verbose", "version"])
    except getopt.GetoptError:
        usage()
        sys.exit(2)
    for opt, arg in opts:
        if opt in ["-h", "--help"]:
            usage()
            sys.exit()
        elif opt in ["-a", "--auth-type"]:
            occi_config["authtype"] = arg
        elif opt in ["-c", "--cert"]:
            occi_config["cert"] = arg
        elif opt in ["-e", "--endpoint", "--url"]:
            occi_config["url"] = arg
        elif opt in ["-f", "--format"]:
            occi_config['outputformat'] = arg
        elif opt in ["-k", "--key"]:
            occi_config["key"] = arg
        elif opt in ["-l", "--list"]:
            print '\n'.join(sorted(tests_definitions.keys()))
            sys.exit()
        elif opt in ["-m", "--mime-type"]:
            occi_config['mimetype'] = arg
        elif opt in ["-P", "--passphrase"]:
            occi_config["passphrase"] = arg
        elif opt in ["-p", "--password"]:
            occi_config["passwd"] = arg
        elif opt in ["-s", "--keystone"]:
            occi_config["keystone"] = arg
        elif opt in ["-S", "--ignore-ssl"]:
            occi_config["ignoressl"] = True
        elif opt in ["-t", "--tests"]:
            tests = arg.split(",")
        elif opt in ["-u", "--user"]:
            occi_config["user"] = arg
        elif opt in ["-v", "--verbose"]:
            occi_config['curlverbose'] = True
        elif opt in ["-V", "--version"]:
            print version.__version__
            sys.exit()

    if not tests:
        tests = sorted(list(set(tests_definitions.keys()) - tests_skipped))

    if 'url' not in occi_config or not occi_config['url']:
        print 'OCCI URL required'
        sys.exit(2)

    if occi_config['authtype'] == 'basic':
        if 'user' not in occi_config or not occi_config['user']:
            print 'User and password is required for "basic" authentization type.'
            sys.exit(2)
    elif occi_config['authtype'] == 'x509':
        if 'X509_USER_PROXY' in os.environ:
            if 'cert' not in occi_config or not occi_config['cert']:
                occi_config['cert'] = os.environ['X509_USER_PROXY']
            if 'cachain' not in occi_config or not occi_config['cachain']:
                occi_config['cachain'] = os.environ['X509_USER_PROXY']
        if 'cert' not in occi_config or not occi_config['cert']:
            print 'SSL certificate and key is required for "x509" authentization type.'
            sys.exit(2)
        if 'capath' not in occi_config or not occi_config['capath']:
            if 'X509_CERT_DIR' in os.environ:
                occi_config['capath'] = os.environ['X509_CERT_DIR']
            elif os.path.isdir('/etc/grid-security/certificates'):
                occi_config['capath'] = '/etc/grid-security/certificates'


    occi_init()
    testsuite_init()

    for test in tests:
        if not re.match(r'OCCI/', test):
            test = 'OCCI/' + test
        if test in tests_definitions.keys():
            start_time = time.time()
            try:
                result, err_msg = tests_definitions[test].test()
            except occi.Error as oe:
                if occi_config['curlverbose']:
                    raise oe
                else:
                    print '%s error: %s' % (test, str(oe))
                    sys.exit(2)
            objective = tests_definitions[test].objective
            running_time = time.time() - start_time
            results.append(occi_test(test, objective, result, err_msg, running_time))

    results = occi_format(results)
    occi_print(results, occi_config['outputformat'])

    if results['failed'] == 0:
        sys.exit(0)
    else:
        sys.exit(1)


if __name__ == "__main__":
    main(sys.argv[1:])
